### PR TITLE
Add retry for Jenkins triggered builds

### PR DESCRIPTION
**Pull request by michaelyoungstrom**
This PR allows the webhook_processor to query the Jenkins REST API, and verify that certain PR's accurately trigger all of their respective Jenkins jobs. If some are left out, it is able to resend the webhook up to a desired number of times.

### CODE REVIEW

File: lambdas/webhook_processor/webhook_processor.py [hunk 1]
```diff
@@ -127,6 +179,224 @@ def _process_results_for_failures(results):
...
+            sha = payload['pull_request']['head']['sha']
+    else:
+        # unsupported event type, return None for both values
+        ignore = True
+
+    # if we are ignoring this hook, return None values,
+    # otherwise, return sha, jobs_list
+    if ignore:
+        # we don't care about these so assign None, None
+        sha, jobs_list = (None, None)
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: nit: it's slightly more readable if you simply return (None, None) here and get rid of the "else"

File: lambdas/webhook_processor/webhook_processor.py [hunk 2]
```diff
@@ -127,6 +179,224 @@ def _process_results_for_failures(results):
...
+        is_merge = payload['pull_request']['merged']
+
+        # check if the base branch is an openedx release branch
+        base_ref = payload['pull_request']['base']['ref']
+        if base_ref == EUCALYTPUS_BRANCH:
+            target = "eucalyptus"
+        elif base_ref == FICUS_BRANCH:
+            target = "ficus"
+        else:
+            target = "master"
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: Do we run the "PR suite" of tests for PRs from any-branch to another-branch-that-isnt-master? Is that why any other value for the base ref should kick off those tests?
Pls add a comment to the code explaining.

File: lambdas/webhook_processor/webhook_processor.py [hunk 3]
```diff
@@ -127,6 +179,224 @@ def _process_results_for_failures(results):
...
+
+        if payload['action'] == 'closed':
+            sha = payload['pull_request']['merge_commit_sha']
+            if not is_merge:
+                # if the PR was closed and not merged,
+                # no jobs will be triggered
+                ignore = True
+        else:
+            # PR was either opened or "synchronized"
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: Pls add a note explaining that "synchronized" is the action name for pushing a new commit (? if that's true. or if not whatever the explanation really is.)

File: lambdas/webhook_processor/webhook_processor.py [hunk 4]
```diff
@@ -127,6 +179,224 @@ def _process_results_for_failures(results):
...
+                    jobs_list = EDX_PLATFORM_FICUS_MASTER
+                else:
+                    jobs_list = EDX_PLATFORM_FICUS_PR
+    elif repository == 'edx-platform-private':
+        if event_type == 'pull_request':
+            if is_merge:
+                jobs_list = EDX_PLATFORM_PRIVATE_MASTER
+            else:
+                jobs_list = EDX_PLATFORM_PRIVATE_PR
+    elif repository == 'edx-e2e-tests':
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: Oh, I don't think that we have this turned on for the edx-e2e-tests repo yet. Can/should we? Let's put in a 1 point JIRA ticket for this (for tracking purposes) and get it in the next sprint if we can.

File: lambdas/webhook_processor/webhook_processor.py [hunk 5]
```diff
@@ -127,6 +179,224 @@ def _process_results_for_failures(results):
...
+    build_status = 'queued'
+
+    # Use Jenkins REST API to get info on the queue
+    # set a timeout for the request to avoid timing out of lambda
+    url = '%s/queue/api/json?depth=0' % (jenkins_url)
+    try:
+        response = get(
+            url,
+            auth=(jenkins_username, jenkins_token),
+            timeout=3
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: nit: you can actually define timeout in a tuple of the connect and read timeouts. Let's keep a pattern of setting the connect timeout to 3.05 and the read timeout to whatever was defined by the environment variable.
Reference: http://docs.python-requests.org/en/master/user/advanced/#timeouts

File: lambdas/webhook_processor/webhook_processor.py [hunk 6]
```diff
@@ -164,12 +434,73 @@ def lambda_handler(event, _context):
...
+        # Get the sha and jobs list (if applicable), as well as
+        # the max number of retries for sending the hook.
+        event_type = headers.get('X-GitHub-Event')
+        sha, jobs_list = _parse_hook_for_testing_info(payload, event_type)
+        num_tries = _get_num_tries()
+
         # Send it off!
-        # Save up the results for later processing rather than letting
-        # the errors get raised.
+        # If the url is a known jenkins instance (in constants.py),
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: We only want to check against build jenkins tho, no? We don't kick off jobs on test-jenkins for platform.

File: lambdas/webhook_processor/webhook_processor.py [hunk 7]
```diff
@@ -34,6 +38,55 @@ def _get_target_urls():
...
+    bucket_name = os.environ.get('S3_CREDENTIALS_BUCKET')
+
+    if not bucket_name:
+        raise StandardError(
+            'Environment variable S3_CREDENTIALS_BUCKET was not set'
+        )
+
+    file_name = JENKINS_S3_OBJECTS[jenkins_url] + '.json'
+
+    creds_file = client.get_object(Bucket=bucket_name, Key=file_name)
```
Comment (jzoldak) [inline lambdas/webhook_processor/webhook_processor.py]: What will happen if the file doesn't exist?

File: lambdas/webhook_processor/webhook_processor.py (hunk 5 above)
Comment (michaelyoungstrom) [inline lambdas/webhook_processor/webhook_processor.py]: which environment variable are you referring to? I see it currently set to (3.05, 10) inside _send_message

File: lambdas/webhook_processor/webhook_processor.py (hunk 2 above)
Comment (michaelyoungstrom) [inline lambdas/webhook_processor/webhook_processor.py]: Yes we do. The jobs do a negative reg ex on 'open-release/' to determine if master PR jobs  should be kicked off. I'll comment.

File: lambdas/webhook_processor/webhook_processor.py (hunk 6 above)
Comment (michaelyoungstrom) [inline lambdas/webhook_processor/webhook_processor.py]: I have removed the test jenkins url from JENKINS_S3_OBJECTS in constants.py . Doing this means the lambda_handler will no longer check for jobs running on test jenkins. 

If it only makes sense to check build jenkins, maybe I should just remove this dict from the constants file, and hard code the url/ file name respectively. Thoughts?

**Pull request closed without merging**